dot/parachain/spam_slots.go

```diff
@@ -0,0 +1,191 @@
+}
+
+func (s *spamSlots) clearSlots(session parachain.SessionIndex, validator parachain.ValidatorIndex) {
+	s.slotsLock.Lock()
+	defer s.slotsLock.Unlock()
+	delete(s.slots, newSlotsKey(session, validator))
+}
+
+func (s *spamSlots) incrementSpamCount(session parachain.SessionIndex, validator parachain.ValidatorIndex) {
+	s.slotsLock.Lock()
+	defer s.slotsLock.Unlock()
+	s.slots[newSlotsKey(session, validator)]++
+}
+
+func (s *spamSlots) AddUnconfirmed(session parachain.SessionIndex, candidate common.Hash, validator parachain.ValidatorIndex) bool {
+	if spamCount, _ := s.getSpamCount(session, validator); spamCount >= s.maxSpamVotes {
+		return false
+	}
+
+	s.unconfirmedLock.Lock()
+	defer s.unconfirmedLock.Unlock()
+
+	validators, ok := s.unconfirmed[newUnconfirmedDisputesKey(session, candidate)]
+	if !ok || validators == nil {
+		validators = treeset.NewWith(byValidatorIndex)
+	}
+
+	if !validators.Contains(validator) {
+		validators.Add(validator)
+		s.incrementSpamCount(session, validator)
+	}
+
+	s.unconfirmed[newUnconfirmedDisputesKey(session, candidate)] = validators
+
+	return true
+}
+
+func (s *spamSlots) Clear(session parachain.SessionIndex, candidate common.Hash) {
+	validators, ok := s.getValidators(session, candidate)
+	if !ok {
+		return
+	}
+
+	validatorSet := validators.Values()
+	s.clearValidators(session, candidate)
+
+	for _, validator := range validatorSet {
+		spamCount, ok := s.getSpamCount(session, validator.(parachain.ValidatorIndex))
+		if ok {
+			if spamCount == 1 {
+				s.clearSlots(session, validator.(parachain.ValidatorIndex))
+				continue
+			}
+
+			s.slotsLock.Lock()
+			s.slots[newSlotsKey(session, validator.(parachain.ValidatorIndex))] = spamCount - 1
+			s.slotsLock.Unlock()
+		}
+	}
+
+}
+
+func (s *spamSlots) PruneOld(oldestIndex parachain.SessionIndex) {
+	s.unconfirmedLock.Lock()
+	maps.DeleteFunc(s.unconfirmed, func(k unconfirmedKey, v *treeset.Set) bool {
+		return k.session < oldestIndex
+	})
+	s.unconfirmedLock.Unlock()
+
+	s.slotsLock.Lock()
+	maps.DeleteFunc(s.slots, func(k slotKey, v uint32) bool {
+		return k.session < oldestIndex
+	})
+	s.slotsLock.Unlock()
+}
+
+// NewSpamSlots returns a new SpamSlots instance
+func NewSpamSlots(maxSpamVotes uint32) SpamSlots {
+	return &spamSlots{
+		slots:        make(map[slotKey]uint32),
+		unconfirmed:  make(map[unconfirmedKey]*treeset.Set),
+		maxSpamVotes: maxSpamVotes,
+	}
+}
+
+// NewSpamSlotsFromState returns a new SpamSlots instance from the given state
+func NewSpamSlotsFromState(unconfirmedDisputes map[unconfirmedKey]*treeset.Set, maxSpamVotes uint32) SpamSlots {
+	slots := make(map[slotKey]uint32)
+
+	for k, v := range unconfirmedDisputes {
+		for validator := range v.Values() {
+			// increment the spam count for this validator and session
+			key := newSlotsKey(k.session, parachain.ValidatorIndex(validator))
+			slots[key]++
+			if slots[key] > maxSpamVotes {
+				// TODO: log this
```

Comment:
this TODO will be addressed in this PR?

Reply:
I'm hoping to. I need to look into the loggers, dunno how it is configured in the project yet.

Reply:
Looks like I am going to skip this for now. This is just going into my feature branch, so I believe it's okay for now. Thank you for the review @EclesioMeloJunior !